dbms/src/Storages/DeltaMerge/SegmentReadTask.cpp

```diff
@@ -214,13 +223,77 @@ void SegmentReadTask::initColumnFileDataProvider(const Remote::RNLocalPageCacheG
         KeyspaceTableID{dm_context->keyspace_id, dm_context->physical_table_id});
 }
 
+
 void SegmentReadTask::initInputStream(
     const ColumnDefines & columns_to_read,
     UInt64 read_tso,
     const PushDownFilterPtr & push_down_filter,
-    ReadMode read_mode)
+    ReadMode read_mode,
+    size_t expected_block_size,
+    bool enable_delta_index_error_fallback)
+{
+    if (likely(doInitInputStream(
+            columns_to_read,
+            read_tso,
+            push_down_filter,
+            read_mode,
+            expected_block_size,
+            enable_delta_index_error_fallback)))
+    {
+        return;
+    }
+
+    // Exception DT_DELTA_INDEX_ERROR raised. Reset delta index and try again.
+    DeltaIndex empty_delta_index;
+    read_snapshot->delta->getSharedDeltaIndex()->swap(empty_delta_index);
+    if (auto cache = dm_context->db_context.getSharedContextDisagg()->rn_delta_index_cache; cache)
+    {
+        cache->setDeltaIndex(read_snapshot->delta->getSharedDeltaIndex());
+    }
+    doInitInputStream(columns_to_read, read_tso, push_down_filter, read_mode, expected_block_size);
+}
+
+bool SegmentReadTask::doInitInputStream(
+    const ColumnDefines & columns_to_read,
+    UInt64 read_tso,
+    const PushDownFilterPtr & push_down_filter,
+    ReadMode read_mode,
+    size_t expected_block_size,
+    bool enable_delta_index_error_fallback)
+{
```

Comment:
This function has the same name as line 283, which is confusing.
Better rename it to `SegmentReadTask::tryInitInputStreamWithFallback(...`